kubeadm re-reads kubeconfig in init phase bootstrap-token

### What happened?

I ran `kubeadm` with an in-memory kubeconfig literal (here reproduced with a file on disk and piped in, the principle is the same) and while many commands worked just fine, this init phase ended in a panic.

```
$ cat admin.conf | ./kubeadm --kubeconfig /dev/stdin init phase bootstrap-token
W1127 10:36:04.466512   80447 version.go:121] could not obtain client version; using remote version: v1.31.3
[bootstrap-token] Using token: vbzyo5.w4wxyalp15a9m0xc
[bootstrap-token] Configuring bootstrap tokens, cluster-info ConfigMap, RBAC Roles
[bootstrap-token] Configured RBAC rules to allow Node Bootstrap tokens to get nodes
[bootstrap-token] Configured RBAC rules to allow Node Bootstrap tokens to post CSRs in order for nodes to get long term certificate credentials
[bootstrap-token] Configured RBAC rules to allow the csrapprover controller automatically approve CSRs from a Node Bootstrap Token
[bootstrap-token] Configured RBAC rules to allow certificate rotation for all node client certificates in the cluster
[bootstrap-token] Creating the "cluster-info" ConfigMap in the "kube-public" namespace
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x10 pc=0x102534c08]

goroutine 1 [running]:
k8s.io/kubernetes/cmd/kubeadm/app/phases/bootstraptoken/clusterinfo.CreateBootstrapConfigMapIfNotExists({0x103107478, 0x14000439a40}, {0x16f8ead0f, 0xa})
	/Users/okokes/git/kubernetes/cmd/kubeadm/app/phases/bootstraptoken/clusterinfo/clusterinfo.go:56 +0x138
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/init.runBootstrapToken({0x102f67ac0?, 0x1400013d220})
	/Users/okokes/git/kubernetes/cmd/kubeadm/app/cmd/phases/init/bootstraptoken.go:109 +0x1f4
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).Run.func1(0x1400022a000)
	/Users/okokes/git/kubernetes/cmd/kubeadm/app/cmd/phases/workflow/runner.go:261 +0xf4
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).visitAll(...)
	/Users/okokes/git/kubernetes/cmd/kubeadm/app/cmd/phases/workflow/runner.go:450
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).Run(0x1400049ed80, {0x1400066a200, 0x0, 0x2})
	/Users/okokes/git/kubernetes/cmd/kubeadm/app/cmd/phases/workflow/runner.go:234 +0x2ac
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).BindToCommand.func1.1(0x140006ad000?, {0x1400066a200, 0x0, 0x2})
	/Users/okokes/git/kubernetes/cmd/kubeadm/app/cmd/phases/workflow/runner.go:376 +0xe0
github.com/spf13/cobra.(*Command).execute(0x140006b0008, {0x1400066a1e0, 0x2, 0x2})
	/Users/okokes/git/kubernetes/vendor/github.com/spf13/cobra/command.go:985 +0x834
github.com/spf13/cobra.(*Command).ExecuteC(0x1400043c608)
	/Users/okokes/git/kubernetes/vendor/github.com/spf13/cobra/command.go:1117 +0x344
github.com/spf13/cobra.(*Command).Execute(0x10309f078?)
	/Users/okokes/git/kubernetes/vendor/github.com/spf13/cobra/command.go:1041 +0x1c
k8s.io/kubernetes/cmd/kubeadm/app.Run()
	/Users/okokes/git/kubernetes/cmd/kubeadm/app/kubeadm.go:47 +0xac
main.main()
	/Users/okokes/git/kubernetes/cmd/kubeadm/kubeadm.go:25 +0x1c
```

### What did you expect to happen?

I expected the command to run just fine.

### How can we reproduce it (as minimally and precisely as possible)?

With a kubeconfig in `admin.conf`, run this in 9d62330bfa31a4fce28093d052f65ff0e88ac3a0:

```
go build -o kubeadm ./cmd/kubeadm
cat admin.conf | ./kubeadm --kubeconfig /dev/stdin init phase bootstrap-token
```


### Anything else we need to know?

I added a debug line to `LoadFromFile` and indeed it gets called twice, so if the source is not to be read repeatedly (e.g. a pipe), it will fail the second time around, because the resulting kubeconfig will be empty

```
$ cat admin.conf | ./kubeadm --kubeconfig /dev/stdin init phase bootstrap-token
W1127 10:12:46.763090   75623 version.go:121] could not obtain client version; using remote version: v1.31.3
2024/11/27 10:12:46 DBG: loading from file: /dev/stdin
[bootstrap-token] Using token: (...)
[bootstrap-token] Configuring bootstrap tokens, cluster-info ConfigMap, RBAC Roles
[bootstrap-token] Configured RBAC rules to allow Node Bootstrap tokens to get nodes
[bootstrap-token] Configured RBAC rules to allow Node Bootstrap tokens to post CSRs in order for nodes to get long term certificate credentials
[bootstrap-token] Configured RBAC rules to allow the csrapprover controller automatically approve CSRs from a Node Bootstrap Token
[bootstrap-token] Configured RBAC rules to allow certificate rotation for all node client certificates in the cluster
[bootstrap-token] Creating the "cluster-info" ConfigMap in the "kube-public" namespace
2024/11/27 10:12:50 DBG: loading from file: /dev/stdin
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x10 pc=0x102bbcb88]

goroutine 1 [running]:
k8s.io/kubernetes/cmd/kubeadm/app/phases/bootstraptoken/clusterinfo.CreateBootstrapConfigMapIfNotExists({0x10378f478, 0x1400062e8c0}, {0x16f266d0f, 0xa})
	/Users/okokes/git/kubernetes/cmd/kubeadm/app/phases/bootstraptoken/clusterinfo/clusterinfo.go:56 +0x138
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/init.runBootstrapToken({0x1035efac0?, 0x14000431c20})
	/Users/okokes/git/kubernetes/cmd/kubeadm/app/cmd/phases/init/bootstraptoken.go:109 +0x1f4
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).Run.func1(0x1400016efc0)
	/Users/okokes/git/kubernetes/cmd/kubeadm/app/cmd/phases/workflow/runner.go:261 +0xf4
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).visitAll(...)
	/Users/okokes/git/kubernetes/cmd/kubeadm/app/cmd/phases/workflow/runner.go:450
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).Run(0x140003a4d80, {0x140004ab860, 0x0, 0x2})
	/Users/okokes/git/kubernetes/cmd/kubeadm/app/cmd/phases/workflow/runner.go:234 +0x2ac
k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow.(*Runner).BindToCommand.func1.1(0x140006cb000?, {0x140004ab860, 0x0, 0x2})
	/Users/okokes/git/kubernetes/cmd/kubeadm/app/cmd/phases/workflow/runner.go:376 +0xe0
github.com/spf13/cobra.(*Command).execute(0x140006da308, {0x140004ab840, 0x2, 0x2})
	/Users/okokes/git/kubernetes/vendor/github.com/spf13/cobra/command.go:985 +0x834
github.com/spf13/cobra.(*Command).ExecuteC(0x14000404908)
	/Users/okokes/git/kubernetes/vendor/github.com/spf13/cobra/command.go:1117 +0x344
github.com/spf13/cobra.(*Command).Execute(0x103727078?)
	/Users/okokes/git/kubernetes/vendor/github.com/spf13/cobra/command.go:1041 +0x1c
k8s.io/kubernetes/cmd/kubeadm/app.Run()
	/Users/okokes/git/kubernetes/cmd/kubeadm/app/kubeadm.go:47 +0xac
main.main()
	/Users/okokes/git/kubernetes/cmd/kubeadm/kubeadm.go:25 +0x1c
```

In the case of this kubeadm phase, it gets read first in `client, err := data.Client()`, but then the absolute path is retrieved and passed to `clusterinfophase.CreateBootstrapConfigMapIfNotExists(client, data.KubeConfigPath())`, where it gets read again.

### Kubernetes version

9d62330bfa31a4fce28093d052f65ff0e88ac3a0

### Cloud provider

<details>

</details>


### OS version

<details>

```console
Darwin foo-bar 23.6.0 Darwin Kernel Version 23.6.0: Wed Jul 31 20:53:05 PDT 2024; root:xnu-10063.141.1.700.5~1/RELEASE_ARM64_T8112 arm64
```
</details>


### Install tools

<details>

</details>


### Container runtime (CRI) and version (if applicable)

<details>

</details>


### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

--- a/cmd/kubeadm/app/cmd/init.go
+++ b/cmd/kubeadm/app/cmd/init.go
@@ -28,6 +28,8 @@ import (
 
 	"k8s.io/apimachinery/pkg/util/sets"
 	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
 	"k8s.io/klog/v2"
 
 	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
@@ -87,6 +89,7 @@ type initData struct {
 	cfg                         *kubeadmapi.InitConfiguration
 	skipTokenPrint              bool
 	dryRun                      bool
+	kubeconfig                  *clientcmdapi.Config
 	kubeconfigDir               string
 	kubeconfigPath              string
 	ignorePreflightErrors       sets.Set[string]
@@ -456,6 +459,21 @@ func (d *initData) CertificateDir() string {
 	return d.certificatesDir
 }
 
+// KubeConfig returns a kubeconfig after loading it from KubeConfigPath().
+func (d *initData) KubeConfig() (*clientcmdapi.Config, error) {
+	if d.kubeconfig != nil {
+		return d.kubeconfig, nil
+	}
+
+	var err error
+	d.kubeconfig, err = clientcmd.LoadFromFile(d.KubeConfigPath())
+	if err != nil {
+		return nil, err
+	}
+
+	return d.kubeconfig, nil
+}
+
 // KubeConfigDir returns the path of the Kubernetes configuration folder or the temporary folder path in case of DryRun.
 func (d *initData) KubeConfigDir() string {
 	if d.dryRun {
@@ -536,7 +554,11 @@ func (d *initData) Client() (clientset.Interface, error) {
 				d.adminKubeConfigBootstrapped = true
 			} else {
 				// Alternatively, just load the config pointed at the --kubeconfig path
-				d.client, err = kubeconfigutil.ClientSetFromFile(d.KubeConfigPath())
+				cfg, err := d.KubeConfig()
+				if err != nil {
+					return nil, err
+				}
+				d.client, err = kubeconfigutil.ToClientSet(cfg)
 				if err != nil {
 					return nil, err
 				}

--- a/cmd/kubeadm/app/cmd/phases/init/bootstraptoken.go
+++ b/cmd/kubeadm/app/cmd/phases/init/bootstraptoken.go
@@ -72,6 +72,10 @@ func runBootstrapToken(c workflow.RunData) error {
 	if err != nil {
 		return err
 	}
+	kubeconfig, err := data.KubeConfig()
+	if err != nil {
+		return err
+	}
 
 	if !data.SkipTokenPrint() {
 		tokens := data.Tokens()
@@ -106,7 +110,7 @@ func runBootstrapToken(c workflow.RunData) error {
 	}
 
 	// Create the cluster-info ConfigMap with the associated RBAC rules
-	if err := clusterinfophase.CreateBootstrapConfigMapIfNotExists(client, data.KubeConfigPath()); err != nil {
+	if err := clusterinfophase.CreateBootstrapConfigMapIfNotExists(client, kubeconfig); err != nil {
 		return errors.Wrap(err, "error creating bootstrap ConfigMap")
 	}
 	if err := clusterinfophase.CreateClusterInfoRBACRules(client); err != nil {

--- a/cmd/kubeadm/app/cmd/phases/init/data.go
+++ b/cmd/kubeadm/app/cmd/phases/init/data.go
@@ -22,6 +22,7 @@ import (
 	"k8s.io/apimachinery/pkg/util/sets"
 	clientset "k8s.io/client-go/kubernetes"
 
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
 	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
 )
 
@@ -38,6 +39,7 @@ type InitData interface {
 	IgnorePreflightErrors() sets.Set[string]
 	CertificateWriteDir() string
 	CertificateDir() string
+	KubeConfig() (*clientcmdapi.Config, error)
 	KubeConfigDir() string
 	KubeConfigPath() string
 	ManifestDir() string

--- a/cmd/kubeadm/app/cmd/phases/init/data_test.go
+++ b/cmd/kubeadm/app/cmd/phases/init/data_test.go
@@ -22,6 +22,7 @@ import (
 	"k8s.io/apimachinery/pkg/util/sets"
 	clientset "k8s.io/client-go/kubernetes"
 
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
 	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
 )
 
@@ -41,6 +42,7 @@ func (t *testInitData) SkipTokenPrint() bool                                 { r
 func (t *testInitData) IgnorePreflightErrors() sets.Set[string]              { return nil }
 func (t *testInitData) CertificateWriteDir() string                          { return "" }
 func (t *testInitData) CertificateDir() string                               { return "" }
+func (t *testInitData) KubeConfig() (*clientcmdapi.Config, error)            { return nil, nil }
 func (t *testInitData) KubeConfigDir() string                                { return "" }
 func (t *testInitData) KubeConfigPath() string                               { return "" }
 func (t *testInitData) ManifestDir() string                                  { return "" }

--- a/cmd/kubeadm/app/phases/bootstraptoken/clusterinfo/clusterinfo.go
+++ b/cmd/kubeadm/app/phases/bootstraptoken/clusterinfo/clusterinfo.go
@@ -21,7 +21,7 @@ import (
 
 	"github.com/pkg/errors"
 
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
 	rbac "k8s.io/api/rbac/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apiserver/pkg/authentication/user"
@@ -40,17 +40,18 @@ const (
 )
 
 // CreateBootstrapConfigMapIfNotExists creates the kube-public ConfigMap if it doesn't exist already
-func CreateBootstrapConfigMapIfNotExists(client clientset.Interface, file string) error {
+func CreateBootstrapConfigMapIfNotExists(client clientset.Interface, kubeconfig *clientcmdapi.Config) error {
 
 	fmt.Printf("[bootstrap-token] Creating the %q ConfigMap in the %q namespace\n", bootstrapapi.ConfigMapClusterInfo, metav1.NamespacePublic)
 
-	klog.V(1).Infoln("[bootstrap-token] loading admin kubeconfig")
-	adminConfig, err := clientcmd.LoadFromFile(file)
-	if err != nil {
-		return errors.Wrap(err, "failed to load admin kubeconfig")
-	}
-	if err = clientcmdapi.FlattenConfig(adminConfig); err != nil {
+	// Clone the kubeconfig so that it's not mutated.
+	adminConfig := kubeconfig.DeepCopy()
+	if err := clientcmdapi.FlattenConfig(adminConfig); err != nil {
 		return err
+	}
+
+	if adminConfig.Contexts[adminConfig.CurrentContext] == nil {
+		return errors.New("invalid kubeconfig")
 	}
 
 	adminCluster := adminConfig.Contexts[adminConfig.CurrentContext].Cluster

--- a/cmd/kubeadm/app/phases/bootstraptoken/clusterinfo/clusterinfo_test.go
+++ b/cmd/kubeadm/app/phases/bootstraptoken/clusterinfo/clusterinfo_test.go
@@ -17,8 +17,8 @@ limitations under the License.
 package clusterinfo
 
 import (
+	"bytes"
 	"context"
-	"os"
 	"testing"
 	"text/template"
 	"time"
@@ -31,6 +31,7 @@ import (
 	"k8s.io/apiserver/pkg/authentication/user"
 	clientsetfake "k8s.io/client-go/kubernetes/fake"
 	core "k8s.io/client-go/testing"
+	"k8s.io/client-go/tools/clientcmd"
 	bootstrapapi "k8s.io/cluster-bootstrap/token/api"
 
 	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
@@ -55,32 +56,22 @@ users:
 func TestCreateBootstrapConfigMapIfNotExists(t *testing.T) {
 	tests := []struct {
 		name      string
-		fileExist bool
 		createErr error
 		expectErr bool
 	}{
 		{
 			"successful case should have no error",
-			true,
 			nil,
 			false,
 		},
 		{
 			"if configmap already exists, return error",
-			true,
 			apierrors.NewAlreadyExists(schema.GroupResource{Resource: "configmaps"}, "test"),
 			true,
 		},
 		{
 			"unexpected error should be returned",
-			true,
 			apierrors.NewUnauthorized("go away!"),
-			true,
-		},
-		{
-			"if the file does not exist, return error",
-			false,
-			nil,
 			true,
 		},
 	}
@@ -93,18 +84,10 @@ func TestCreateBootstrapConfigMapIfNotExists(t *testing.T) {
 	}
 
 	for _, server := range servers {
-		file, err := os.CreateTemp("", "")
-		if err != nil {
-			t.Fatalf("could not create tempfile: %v", err)
-		}
-		defer os.Remove(file.Name())
+		var buf bytes.Buffer
 
-		if err := testConfigTempl.Execute(file, server); err != nil {
+		if err := testConfigTempl.Execute(&buf, server); err != nil {
 			t.Fatalf("could not write to tempfile: %v", err)
-		}
-
-		if err := file.Close(); err != nil {
-			t.Fatalf("could not close tempfile: %v", err)
 		}
 
 		// Override the default timeouts to be shorter
@@ -124,11 +107,12 @@ func TestCreateBootstrapConfigMapIfNotExists(t *testing.T) {
 					})
 				}
 
-				fileName := file.Name()
-				if !tc.fileExist {
-					fileName = "notexistfile"
+				kubeconfig, err := clientcmd.Load(buf.Bytes())
+				if err != nil {
+					t.Fatal(err)
 				}
-				err := CreateBootstrapConfigMapIfNotExists(client, fileName)
+
+				err = CreateBootstrapConfigMapIfNotExists(client, kubeconfig)
 				if tc.expectErr && err == nil {
 					t.Errorf("CreateBootstrapConfigMapIfNotExists(%s) wanted error, got nil", tc.name)
 				} else if !tc.expectErr && err != nil {

